File: crates/sys/src/types/tree.rs
use serde::{Deserialize, Serialize};

use crate::{cast_ptr, convert_ptr_to_vec, get_string_from_ptr};

#[derive(Debug, Serialize, Deserialize)]
pub struct TreeCache {
    pub entries_handle: i32,     // int entries_handle
    pub name_buffer_handle: i32, // int name_buffer_handle
    pub max_entries: i32,        // int max_entries
    pub name_buffer_size: i32,   // int name_buffer_size (in bytes)
}

impl From<crate::TreeCache> for TreeCache {
    fn from(cache: crate::TreeCache) -> Self {
        Self {
            entries_handle: cache.entries_handle,
            name_buffer_handle: cache.name_buffer_handle,
            max_entries: cache.max_entries,
            name_buffer_size: cache.name_buffer_size,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TreeContext {
    pub currdir: String,                 // char currdir[MAX_PATH]
    pub dirlevel: i32,                   // int dirlevel
    pub selected_item: i32,              // int selected_item
    pub selected_item_history: Vec<i32>, // int selected_item_history[MAX_DIR_LEVELS]
    pub filesindir: i32,                 // int filesindir
    pub dirsindir: i32,                  // int dirsindir
    pub dirlength: i32,                  // int dirlength
    pub currtable: i32,                  // int currtable (db use)
    pub currextra: i32,                  // int currextra (db use)
    pub sort_dir: i32,                   // int sort_dir
    pub out_of_tree: i32,                // int out_of_tree
    pub cache: TreeCache,                // struct tree_cache cache
    pub dirfull: bool,                   // bool dirfull
    pub is_browsing: bool,               // bool is_browsing
    pub browse: Option<BrowseContext>,   // struct browse_context* browse
}

impl From<crate::TreeContext> for TreeContext {
    fn from(context: crate::TreeContext) -> Self {
        Self {
            currdir: unsafe {
                std::ffi::CStr::from_ptr(cast_ptr!(context.currdir.as_ptr()))
                    .to_string_lossy()
                    .into_owned()
            },
            dirlevel: context.dirlevel,
            selected_item: context.selected_item,
            selected_item_history: context.selected_item_history.to_vec(),
            filesindir: context.filesindir,
            dirsindir: context.dirsindir,
            dirlength: context.dirlength,
            currtable: context.currtable,
            currextra: context.currextra,
            sort_dir: context.sort_dir,
            out_of_tree: context.out_of_tree,
            cache: context.cache.into(),
            dirfull: context.dirfull,
            is_browsing: context.is_browsing,
            browse: None,
            // browse: ptr_to_option!(context.browse).map(|browse| browse.into()),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BrowseContext {
    pub dirfilter: i32, // int dirfilter
    pub flags: u32,     // unsigned flags
    pub title: String,  // char* title
    // pub icon: ThemableIcons, // enum themable_icons icon
    pub root: String,     // const char* root
    pub selected: String, // const char* selected
    pub buf: Vec<i8>,     // char* buf
    pub bufsize: usize,   // size_t bufsize
}

impl From<crate::BrowseContext> for BrowseContext {
    fn from(context: crate::BrowseContext) -> Self {
        Self {
            dirfilter: context.dirfilter,
            flags: context.flags,
            title: get_string_from_ptr!(context.title),
            root: get_string_from_ptr!(context.root),
            selected: get_string_from_ptr!(context.selected),
            buf: convert_ptr_to_vec!(context.buf, context.bufsize),
            bufsize: context.bufsize,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Entry {
    pub name: String,      // char* name
    pub attr: i32,         // int attr (FAT attributes + file type flags)
    pub time_write: u32,   // unsigned time_write (Last write time)
    pub customaction: i32, // int customaction (db use)
}

impl From<crate::Entry> for Entry {
    fn from(entry: crate::Entry) -> Self {
        Self {
            name: get_string_from_ptr!(entry.name),
            attr: entry.attr,
            time_write: entry.time_write,
            customaction: entry.customaction,
        }
    }
}

File: zig/src/rockbox/tree.zig
const std = @import("std");

pub const MAX_PATH = 260;
pub const MAX_DIR_LEVELS = 10;

const themable_icons = enum(c_int) {
    NOICON = -2,
    Icon_NOICON = -1,
    Icon_Audio,
    Icon_Folder,
    Icon_Playlist,
    Icon_Cursor,
    Icon_Wps,
    Icon_Firmware,
    Icon_Font,
    Icon_Language,
    Icon_Config,
    Icon_Plugin,
    Icon_Bookmark,
    Icon_Preset,
    Icon_Queued,
    Icon_Moving,
    Icon_Keyboard,
    Icon_Reverse_Cursor,
    Icon_Questionmark,
    Icon_Menu_setting,
    Icon_Menu_functioncall,
    Icon_Submenu,
    Icon_Submenu_Entered,
    Icon_Recording,
    Icon_Voice,
    Icon_General_settings_menu,
    Icon_System_menu,
    Icon_Playback_menu,
    Icon_Display_menu,
    Icon_Remote_Display_menu,
    Icon_Radio_screen,
    Icon_file_view_menu,
    Icon_EQ,
    Icon_Rockbox,
    Icon_Last_Themeable,
};

pub const browse_context = extern struct {
    dirfilter: c_int,
    flags: c_uint, // 'unsigned' is translated to 'c_uint'
    title: [*c]const u8, // Nullable C string for the title
    icon: themable_icons, // Enum 'themable_icons', assumed to be defined elsewhere
    root: [*c]const u8, // Const C string for root directory path
    selected: [*c]const u8, // Const C string for selected file name
    buf: [*c]u8, // Buffer for the selected file
    bufsize: usize, // Size of the buffer (translated to 'usize' for portability)
};

const tree_cache = extern struct {
    entries_handle: c_int, // Handle to the entry cache
    name_buffer_handle: c_int, // Handle to the name cache
    max_entries: c_int, // Maximum number of entries in the cache
    name_buffer_size: c_int, // Size of the name buffer (in bytes)
};

pub const tree_context = extern struct {
    currdir: [MAX_PATH]u8, // Fixed-size array for the current directory
    dirlevel: i32, // int in C is c_int in Zig
    selected_item: i32, // Selected file/id3dbitem index
    selected_item_history: [MAX_DIR_LEVELS]c_int, // History of selected items, fixed-size array

    dirfilter: ?*c_int, // Nullable pointer to an int for file use
    filesindir: i32, // Number of files in the directory cache
    dirsindir: i32, // Directory use
    dirlength: i32, // Total number of entries in directory

    currtable: i32,
    currextra: i32,

    sort_dir: i32, // Directory sort order
    out_of_tree: i32, // Shortcut from elsewhere
    cache: tree_cache, // Struct tree_cache, defined elsewhere

    dirfull: bool,
    is_browsing: bool,

    browse: ?*browse_context, // Pointer to browse_context, nullable
};
pub const entry = extern struct {
    name: ?*c_char, // Pointer to the name (nullable)
    attr: c_int, // FAT attributes + file type flags
    time_write: c_uint, // Last write time (unsigned)
    customaction: c_int, // Custom action (for database use)
};

extern fn tree_init() void;
extern fn rockbox_browse(browse: ?*browse_context) c_int;
extern fn rockbox_browse_at(path: [*]const u8) c_int;
extern fn tree_get_context() *tree_context;
extern fn tree_get_entries(t: *tree_context) *entry;
extern fn tree_get_entry_at(t: *tree_context, index: c_int) *entry;

pub fn _rockbox_browse() c_int {
    var browse: browse_context = .{
        .dirfilter = 0,
        .flags = 0,
        .title = "demo",
        .icon = themable_icons.Icon_NOICON,
        .root = "/",
        .selected = null,
        .buf = null,
        .bufsize = 0,
    };
    return rockbox_browse(&browse);
}

pub fn _tree_get_context() tree_context {
    const tc = tree_get_context();
    return tc.*;
}

pub fn _tree_get_entries() *entry {
    const tc = tree_get_context();
    return tree_get_entries(tc);
}

pub fn _tree_get_entry_at(index: c_int) entry {
    const tc = tree_get_context();
    const e = tree_get_entry_at(tc, index);
    return e.*;
}
